[SYCL][ESIMD] Add utility function for constructing conversion ref data

vasilytric opened the pull request:
- Implement function for fill vector with provided depends to number values in vector
- Implementing function for obtaining reference converted data

vasilytric committed:
[SYCL][ESIMD] Add utility function for constructing ref data

It removes duplicated code

vasilytric committed:
[SYCL][ESIMD] Add function for generate converted reference data

This function provides reference data in the tests that uses a few data
types.

vasilytric committed:
[SYCL][ESIMD] Rename generate_converted_ref_data function

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -59,6 +59,27 @@ sycl::half half_from_bytes(unsigned char hi, unsigned char lo) {
+    for (size_t i = ref_data.size(); i < NumElems; ++i) {
+      ref_data.push_back(i + step);
+    }
+  }
```
Please, use descriptive names for parameters.
There are two concerns on naming selection for me here:

1. The `step` parameter of `construct_ref_data` is misleading. Looks like the `++i` gives the step for the tail part of the vector. What would you say about something like `frac`, as it defines the fractional part of such elements?
2. The names `first_vector` and `second_vector` don't have any semantical meaning. May be something like `essential_values` and `extra_values`?

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/696/commits/c9bc321af2b1bf1b9c70b323c48041dd1e79a85e#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L64-R66)

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -59,6 +59,27 @@ sycl::half half_from_bytes(unsigned char hi, unsigned char lo) {
+template <typename DataT, int NumElems>
+std::vector<DataT> construct_ref_data(double step,
+                                      std::vector<DataT> first_vector,
+                                      std::vector<DataT> second_vector = {}) {
```
```suggestion
                                      std::vector<DataT>&& first_vector,
                                      const std::vector<DataT>& second_vector = {}) {
```
Looks like you are using rvalues anyway, so you might use move constructor for the `first_vector` part.

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/696/commits/c9bc321af2b1bf1b9c70b323c48041dd1e79a85e#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L64-R66)

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -135,41 +156,76 @@ template <typename DataT, int NumElems> std::vector<DataT> generate_ref_data() {
-      }
-    }
+    ref_data = details::construct_ref_data<DataT, NumElems>(
+        0, {min, min_half, max, max_half, 0},
```
Please note that there is no need for explicit `0` here. We might use the same `0.25` with no effect. Therefore we might use `0.25` as a default value for this parameter.

vasilytric replied:
I've done it in this commit: c9bc321af2b1bf1b9c70b323c48041dd1e79a85e

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -135,41 +156,76 @@ template <typename DataT, int NumElems> std::vector<DataT> generate_ref_data() {
+  static_assert(std::is_integral_v<SrcT> ||
+                    type_traits::is_sycl_floating_point_v<SrcT>,
+                "Invalid the first data type provided to the "
+                "generate_ref_conv_data function as source type.");
```
```suggestion
                "Invalid source type");
```
No need to mention "first" or "second". Also no need to mention the function name. Here and below.

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/696/commits/c9bc321af2b1bf1b9c70b323c48041dd1e79a85e#diff-e031be0968a4e0cad81b5b2879e34920db09777fa3165f4f722fb4e0ab560826R71)

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -135,41 +156,76 @@ template <typename DataT, int NumElems> std::vector<DataT> generate_ref_data() {
+                "generate_ref_conv_data function as destination type.");
+
+  static const SrcT min =
+      std::max(value<SrcT>::lowest(), static_cast<SrcT>(value<DstT>::lowest()));
```
From the first look we might trigger UB by `static_cast<SrcT>(value<DstT>)` with signed integral types. Here and below. We need some comments for each such cast at least, stating the reason why it's valid.

yuriykoch replied:
Looks like we really end up in UB by using `static_cast<std::make_signed_t<T>>(T value)`. C++17 states that the resulting value is unspecified in such case.
May be we might avoid it with some `value_conv<SrctT, DstT>::lowest()` here?

vasilytric replied:
I've implemented the separate file that contained all necessary thangs for , it can be found [here](https://github.com/intel/llvm-test-suite/pull/696/commits/c9bc321af2b1bf1b9c70b323c48041dd1e79a85e#diff-e031be0968a4e0cad81b5b2879e34920db09777fa3165f4f722fb4e0ab560826)

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -135,41 +156,76 @@ template <typename DataT, int NumElems> std::vector<DataT> generate_ref_data() {
+  static const SrcT nan = value<SrcT>::nan();
+  static const SrcT inf = value<SrcT>::inf();
+  static const SrcT denorm = std::max(
+      value<SrcT>::denorm_min(), static_cast<SrcT>(value<DstT>::denorm_min()));
```
What if `SrcT` is not a floating point type?

vasilytric replied:
Will be triggered `static_assert`, I've done in [here](https://github.com/intel/llvm-test-suite/pull/696/commits/c9bc321af2b1bf1b9c70b323c48041dd1e79a85e#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514R111-R113)

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -135,41 +156,76 @@ template <typename DataT, int NumElems> std::vector<DataT> generate_ref_data() {
+      value<SrcT>::denorm_min(), static_cast<SrcT>(value<DstT>::denorm_min()));
+  static const SrcT positive = static_cast<SrcT>(126.75);
+
+  std::vector<SrcT> ref_data{};
```
```suggestion
  std::vector<SrcT> ref_data;
```
Are you aware of any difference for such case?

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/696/commits/c9bc321af2b1bf1b9c70b323c48041dd1e79a85e#diff-e031be0968a4e0cad81b5b2879e34920db09777fa3165f4f722fb4e0ab560826R85)

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -135,41 +156,76 @@ template <typename DataT, int NumElems> std::vector<DataT> generate_ref_data() {
+
+  std::vector<SrcT> ref_data{};
+
+  if constexpr (type_traits::is_sycl_floating_point_v<SrcT> &&
+                type_traits::is_sycl_floating_point_v<DstT>) {
+    ref_data = details::construct_ref_data<SrcT, NumElems>(
+        0.25, {min, max, -0.0, +0.0, 0.1, denorm, nan, -inf});
+  } else if constexpr (type_traits::is_sycl_floating_point_v<SrcT> &&
+                       std::is_unsigned_v<DstT>) {
+    ref_data = details::construct_ref_data<SrcT, NumElems>(
+        0.75, {-0.0, max, max_half, -max_half});
```
So we basically have `0.75`, `0.25` and `0` for the fractional part of reference data tail. Given the use cases we could talk about `0.75` and `0.25` values. The only reason we want `0.75` is to verify that fractional part doesn't affect the conversion to integral type at all, that there is no some "upper bound" rounding. The only reason we have `0.25` for is to have some fractional part, to make sure it preserves.

So what if we just use `0.75` everywhere, with no additional parameter passed to `details::construct_ref_data`? Are there any caveats I don't see?

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/696/commits/c9bc321af2b1bf1b9c70b323c48041dd1e79a85e#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L162-L171)

vasilytric committed:
[SYCL][ESIMD] Update value_conv function

- Move all things that depends to value_conv into separate file
- Add value_conv to savety retreive converted reference value
- Update comments
- Remove uniform initialisation

yuriykoch commented on SYCL/ESIMD/api/functional/value_conv.hpp:
```diff
@@ -0,0 +1,117 @@
+// static_cast<int>(unsigned int).
+template <typename SrcT, typename DstT> struct value_conv {
+  static SrcT lowest() {
+    if constexpr (!sts::is_unsigned_v<SrcT> && sts::is_unsigned_v<DstT>) {
```
```suggestion
    if constexpr (std::is_signed_v<SrcT> && std::is_unsigned_v<DstT>) {
```
A tiny misprint, here and elsewhere. Also why not to use `std::is_signed` instead of `!std::is_unsigned`?

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/696/commits/3ae1bec31097fff5e7647d14e3bac7e89b5f497a#diff-e031be0968a4e0cad81b5b2879e34920db09777fa3165f4f722fb4e0ab560826R28)

yuriykoch commented on SYCL/ESIMD/api/functional/value_conv.hpp:
```diff
@@ -0,0 +1,117 @@
+// All provided methods are save to use and protected from UB when call
+// static_cast<int>(unsigned int).
+template <typename SrcT, typename DstT> struct value_conv {
+  static SrcT lowest() {
```
```suggestion
  static inline SrcT lowest() {
```

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/696/commits/3ae1bec31097fff5e7647d14e3bac7e89b5f497a#diff-e031be0968a4e0cad81b5b2879e34920db09777fa3165f4f722fb4e0ab560826R27)

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -61,16 +61,17 @@ sycl::half half_from_bytes(unsigned char hi, unsigned char lo) {
-                                      std::vector<DataT> first_vector,
-                                      std::vector<DataT> second_vector = {}) {
+std::vector<DataT>
+construct_ref_data(double step = 0.75, std::vector<DataT> &&essential_values,
```
Do you still need the `step` parameter? See suggestion from discussion [above](https://github.com/intel/llvm-test-suite/pull/696#discussion_r781215236)

vasilytric replied:
This parameter has removed [here](https://github.com/intel/llvm-test-suite/pull/696/commits/3ae1bec31097fff5e7647d14e3bac7e89b5f497a#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L65)

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -61,16 +61,17 @@ sycl::half half_from_bytes(unsigned char hi, unsigned char lo) {
+std::vector<DataT>
+construct_ref_data(double step = 0.75, std::vector<DataT> &&essential_values,
+                   const std::vector<DataT> &extra_values = {}) {
   std::vector<DataT> ref_data{};
```
You might just use `std::move(essential_values)` here.

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/696/commits/3ae1bec31097fff5e7647d14e3bac7e89b5f497a#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514R67)

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -107,6 +108,8 @@ template <typename DataT> struct value {
 
   static DataT denorm_min() {
+    static_assert(type_traits::is_sycl_floating_point_v<DataT>,
+                  "Only floating point type can be passed into this function.");
```
The method is not a template itself. Doesn't instantiation of the `value<int>` class trigger this static assert?

vasilytric replied:
`static_assert` was replaced with `assert` [here](url)

yuriykoch commented on SYCL/ESIMD/api/functional/value_conv.hpp:
```diff
@@ -0,0 +1,117 @@
+      return static_cast<SrcT>(value<DstT>::max());
+    } else {
+      return value<SrcT>::max();
+    }
```
```suggestion
    if constexpr (!type_traits::is_sycl_floating_point_v<DataT>) {
      // Return zero for any integral type the same way std::denorm_min does
      return 0;
    } else if constexpr (sizeof(SrcT) > sizeof(DstT)) {
      // Use the biggest value so it would not degenerate to zero during conversion
      // from SrcT to DstT
      return static_cast<SrcT>(value<DstT>::denorm_min());
    } else {
      return value<SrcT>::denorm_min();
    }
```

So, we need to have something like `std::max(src_denorm, dst_denorm)` with UB resolved, right?

1. May be it's better to state explicitly value returned for non-floating point `SrcT` case?
2. Why do we need to return `value<SrcT>::denorm_min()` for the case of `using DstT = sycl::half`? By doing so we return the minimal value out of the DstT and SrcT values.
3. According to SYCL 2020 "The half data type must conform to the IEEE 754-2008 half precision storage format.". It means we don't need the separate branches for the `sycl::half` at all. Am I missing something?

vasilytric replied:
I've applied this comment in this commit: 5959fb1f2f4dd5aca794998f3a625824bfc42107

yuriykoch commented on SYCL/ESIMD/api/functional/value_conv.hpp:
```diff
@@ -0,0 +1,117 @@
+
+  std::vector<SrcT> ref_data;
+
+  if constexpr (type_traits::is_sycl_floating_point_v<SrcT> &&
+                type_traits::is_sycl_floating_point_v<DstT>) {
+    ref_data = details::construct_ref_data<SrcT, NumElems>(
+        {min, max, -0.0, +0.0, 0.1, denorm, nan, -inf});
+  } else if constexpr (type_traits::is_sycl_floating_point_v<SrcT> &&
+                       std::is_unsigned_v<DstT>) {
```
As a suggestion - may be it's better to go with nested `if` conditions? Looks like something like
```
if constexpr (type_traits::is_sycl_floating_point_v<SrcT>) {
   // Provide reference data for conversion from the floating point types

   static const SrcT nan = value<SrcT>::nan();
   static const SrcT inf = value<SrcT>::inf();
   static const SrcT denorm = value_conv<SrcT, DstT>::denorm_min();

   if constexpr (type_traits::is_sycl_floating_point_v<DstT>) {
   } else if constexpr (std::is_unsigned_v<DstT>) {
   } else {
   }
} else
```
would give some nested "scopes" of logic we could visually differentiate.

vasilytric replied:
I suggest do it in separate PR because we need more time to avoid UB cases
I leave the TODO comment [here](https://github.com/intel/llvm-test-suite/pull/696/commits/5959fb1f2f4dd5aca794998f3a625824bfc42107#diff-e031be0968a4e0cad81b5b2879e34920db09777fa3165f4f722fb4e0ab560826R52-R53)

yuriykoch commented on SYCL/ESIMD/api/functional/value_conv.hpp:
```diff
@@ -0,0 +1,117 @@
+      return std::min(value<SrcT>::max(),
+                      static_cast<SrcT>(value<DstT>::max()));
+    }
+  }
```
So, we need something like `std::min<src_max, dst_max>`. 
Looks like the logic should be a bit more complex here to avoid UBs though. Even for concrete values like `max()` or `lowest()`. Just for example: we end up with UB for `using SrcT = char; using DstT = int;`

Maybe we might use helpers with specialization over specific types. I have no good suggestion right now, so let me have some time to consider maybe. Feel free to suggest any solutions you find appropriate though.

vasilytric replied:
I suggest do it in separate PR because we need more time to avoid UB cases
I leave the TODO comment [here](https://github.com/intel/llvm-test-suite/pull/696/commits/5959fb1f2f4dd5aca794998f3a625824bfc42107#diff-e031be0968a4e0cad81b5b2879e34920db09777fa3165f4f722fb4e0ab560826R52-R53)

vasilytric committed:
[SYCL][ESIMD] Small updates

- Use std::move to init std::vector
- Make static functions inlined
- Remove estatic_assert

vasilytric committed:
SYCL][ESIMD] Remove function for obtain lowest and max functions

vasilytric committed:
Merge branch 'intel' into update_value

vasilytric committed:
[SYCL][ESIMD] Remove useless std::move

vasilytric commented:
CI on Windows platform failed due to issue, that is not related to this PR.
@sndmitriev  Would you mind to take a look?

vasilytric commented:
CI on Windows failed due to `Basic/image/image_accessor_readsampler.cpp` test failure, but this test is not related to this PR.
@kbobrovs Could you merge this PR?

kbobrovs commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -59,6 +59,26 @@ sycl::half half_from_bytes(unsigned char hi, unsigned char lo) {
   return esimd_test::bit_cast<sycl::half>(out);
 }
 
+// Constructs std::vector with from provided vectors.
```
```suggestion
// Constructs std::vector from provided vectors.
```
But the description does not reflect the (very complex) function logic - please fix.

vasilytric replied:
I've applied this comment in this commit: 00693ecea35d0a2935c72bb247a7a9835f965a42

kbobrovs commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -86,6 +106,8 @@ template <typename DataT> struct value {
 
   static DataT denorm_min() {
+    assert(type_traits::is_sycl_floating_point_v<DataT> &&
+           "Only floating point type can be passed into this function.");
```
please add comments for the bit values below (what half value they represent)

vasilytric replied:
I've applied this comment in this commit: 00693ecea35d0a2935c72bb247a7a9835f965a42

kbobrovs commented on SYCL/ESIMD/api/functional/value_conv.hpp:
```diff
@@ -0,0 +1,91 @@
+
+// Utility class to retrieve specific values for tests depending on the source
+// and destination data types. May be used to retrieve converted reference data.
+// All provided methods are save to use and protected from UB when call
```
```suggestion
// All provided methods are safe to use and protected from UB when call
```
Not clear what the last sentence actually means, please rephrase,

vasilytric replied:
I've applied this comment in this commit: 00693ecea35d0a2935c72bb247a7a9835f965a42

vasilytric committed:
[SYCL][ESIMD] Update comments
